File: segmentController/src/main/java/com/clumob/segment/controller/SegmentInfo.java
package com.clumob.segment.controller;

import android.os.Parcel;

import com.clumob.segment.controller.util.ParcelableUtil;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;

/**
 * Created by prashant.rathore on 02/02/18.
 */

public class SegmentInfo<Args extends Storable, RestoreableState extends Storable> implements Storable {

    private static final String KEY_ID = "id";
    private static final String KEY_PARAMS = "params";
    private static final String KEY_SAVED_STATE = "savedState";

    int id;
    private Args arguments;
    private RestoreableState restorableSetmentState;

    SegmentInfo() {

    }

    public SegmentInfo(int id, Args args) {
        this.id = id;
        this.arguments = args;
    }

    public Args getArguments() {
        return arguments;
    }

    protected SegmentInfo(Parcel in) {
        id = in.readInt();
        try {
            arguments = readParcel(in);
        } catch (Exception e) {
            e.printStackTrace();
        }
        try {
            restorableSetmentState = readParcel(in);
        } catch (Exception e) {
            e.printStackTrace();
        }

    }

    private <T> T readParcel(Parcel in) throws ClassNotFoundException, IllegalAccessException, InstantiationException, NoSuchMethodException, InvocationTargetException {
        T parcelable = null;
        int readByteArrayLength = in.readInt();
        if (readByteArrayLength > 0) {
            final String className = in.readString();
            byte[] dataArray = new byte[readByteArrayLength];
            in.readByteArray(dataArray);
            Constructor<?> constructor = Class.forName(className).getDeclaredConstructor();
            constructor.setAccessible(true);
            Creator<T> creator = (Creator) constructor.newInstance();
            parcelable = ParcelableUtil.unmarshall(dataArray, creator);
        }
        return parcelable;
    }

    public static final Creator<SegmentInfo> CREATOR = new Creator<SegmentInfo>() {
        @Override
        public SegmentInfo createFromParcel(Parcel in) {
            return new SegmentInfo(in);
        }

        @Override
        public SegmentInfo[] newArray(int size) {
            return new SegmentInfo[size];
        }
    };

    public RestoreableState getRestorableSetmentState() {
        return restorableSetmentState;
    }

    public int getId() {
        return id;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel parcel, int i) {
        parcel.writeInt(id);
        writeToParcel(arguments, parcel);
        writeToParcel(restorableSetmentState, parcel);
    }

    public void setRestorableSegmentState(RestoreableState restorableSetmentState) {
        this.restorableSetmentState = restorableSetmentState;
    }

    private <T extends Storable> void writeToParcel(T storable, Parcel parcel) {
        if (storable != null) {
            Creator<?> creator = storable.creator();
            if(creator == null) {
                throw new NullPointerException("Creator object cannot be null for " + getClass().getName());
            }
            final Class aClass = creator.getClass();
            final String aClassName = aClass.getName();
            byte[] marshall = ParcelableUtil.marshall(storable);
            final int length = marshall.length;
            if (length > 0) {
                parcel.writeInt(length);
                parcel.writeString(aClassName);
                parcel.writeByteArray(marshall);
                return;
            }
        }
        parcel.writeInt(-1);
        return;
    }

    @Override
    public Creator<? extends SegmentInfo> creator() {
        return CREATOR;
    }

}

File: segmentController/src/main/java/com/clumob/segment/controller/SegmentInfo.kt
package com.clumob.segment.controller

import android.os.Parcel
import android.os.Parcelable
import com.clumob.segment.controller.util.ParcelableUtil
import java.lang.reflect.InvocationTargetException

/**
 * Created by prashant.rathore on 02/02/18.
 */
open class SegmentInfo : Storable {

    var id = 0
    var arguments: Storable? = null
        private set
    var restorableSetmentState: Storable? = null
        private set

    internal constructor() {}
    constructor(id: Int, args: Storable?) {
        this.id = id
        arguments = args
    }

    protected constructor(input: Parcel) {
        id = input.readInt()
        try {
            arguments = readParcel<Storable>(input)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        try {
            restorableSetmentState = readParcel<Storable>(input)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    @Throws(ClassNotFoundException::class, IllegalAccessException::class, InstantiationException::class, NoSuchMethodException::class, InvocationTargetException::class)
    private fun <T> readParcel(input: Parcel): T? {
        var parcelable: T? = null
        val readByteArrayLength = input.readInt()
        if (readByteArrayLength > 0) {
            val className = input.readString()
            val dataArray = ByteArray(readByteArrayLength)
            input.readByteArray(dataArray)
            val constructor = Class.forName(className).getDeclaredConstructor()
            constructor.isAccessible = true
            val creator: Parcelable.Creator<T> = constructor.newInstance() as Parcelable.Creator<T>
            parcelable = ParcelableUtil.unmarshall(dataArray, creator)
        }
        return parcelable
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(parcel: Parcel, i: Int) {
        parcel.writeInt(id)
        writeToParcel<Storable?>(arguments, parcel)
        writeToParcel<Storable?>(restorableSetmentState, parcel)
    }

    fun setRestorableSegmentState(restorableSetmentState: Storable?) {
        this.restorableSetmentState = restorableSetmentState
    }

    private fun <T : Storable?> writeToParcel(storable: T?, parcel: Parcel) {
        if (storable != null) {
            val creator = storable.creator()
                    ?: throw NullPointerException("Creator object cannot be null for " + javaClass.name)
            val aClass: Class<*> = creator.javaClass
            val aClassName = aClass.name
            val marshall = ParcelableUtil.marshall(storable)
            val length = marshall.size
            if (length > 0) {
                parcel.writeInt(length)
                parcel.writeString(aClassName)
                parcel.writeByteArray(marshall)
                return
            }
        }
        parcel.writeInt(-1)
    }

    override fun creator(): Parcelable.Creator<out SegmentInfo> {
        return CREATOR
    }

    companion object {
        private const val KEY_ID = "id"
        private const val KEY_PARAMS = "params"
        private const val KEY_SAVED_STATE = "savedState"
        @JvmField
        val CREATOR: Parcelable.Creator<SegmentInfo> = object : Parcelable.Creator<SegmentInfo> {
            override fun createFromParcel(input: Parcel): SegmentInfo {
                return SegmentInfo(input)
            }

            override fun newArray(size: Int): Array<SegmentInfo?> {
                return arrayOfNulls(size)
            }
        }
    }
}